OKTA-375130 Redirect vs Embedded Authentication OIE Concept (#2195)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/concepts/redirect-vs-embedded/index.md

---

> title: Redirect vs Embedded Authentication

Author: We're also not supposed to use Latin, although MMS seems fine with using versus. It works best here, imo. Thanks for the style note. Updated.

meta:
  - name: description
    content: A high-level overview of the the redirect and embedded authentication models for Okta Identity engine.
---

> # Redirect Authentication Vs Embedded Authentication

Author: Yes, we do. Good catch. Thanks.

<ApiLifecycle access="ie" /><br>
<ApiLifecycle access="Limited GA" />

> When developing applications requiring customer sign in and authentication, the deployment model of user authentication is a critical design consideration, which can broadly be divided into two types:

Author: I like approaches. Thanks.

> When developing applications requiring customer sign in and authentication, the deployment model of user authentication is a critical design consideration, which can broadly be divided into two approaches:

Author: This is a high-level comment and I think ", which" works a little better here, i.e., the clause before can stand on its own. But if you think it reads better I can update it. There's no real substantial confusion here using either.

> When developing applications requiring customer sign in and authentication, the deployment model of user authentication is a critical design consideration, which can broadly be divided into two approaches:

Reviewer: okay - so if "which can broadly be divided into two approaches" isn't necessary and the clause before can stand on its own, perhaps just remove it all together for fewer words?

When developing applications that require customer sign in and authentication, the deployment model of user authentication is a critical design consideration:

??

- **Redirect authentication**: A user sign-in flow that grants authentication control to Okta by redirecting to an Okta hosted sign-in page using open protocols like OAuth 2.0 and SAML.

> - **Direct authentication**: A user sign-in flow where the application retains authentication control, without redirection to Okta, using a client-hosted Sign-In Widget, Identity Experience (IDX) SDKs, or directly with Okta’s proprietary IDX authentication APIs.

Author: As per Van's update on Slack, I'll use Identity Engine here instead of Identity Experience.

> The Identity engine supports both these deployment and authentication models, and uses authentication protocols to communicate with the Identity Server, the entity that acts as the logical location for all identity-related services and resources.

Author: We probably need more department guidance on this style. I originally took it to mean API objects, but not sure what constitutes an Okta object, at least from our style perspective. For now, I think I'll capitalize "Identity", as per our Identity engine model.

> The Identity Engine supports both these deployment and authentication models, and uses authentication protocols to communicate with the Identity server, the entity that acts as the logical location for all identity-related services and resources.

Author: Van suggested we lower case "server" here :o) we both had some concerns that we needed to clarify some of our capitalization styles. But I'll go with you here and search-and-replace for other instances.

> The Identity Engine supports both these deployment and authentication models, and uses authentication protocols to communicate with the Identity server, the entity that acts as the logical location for all identity-related services and resources.

Reviewer: > Van suggested we lower case "server" here :o) we both had some concerns that we needed to clarify some of our capitalization styles. But I'll go with you here and search-and-replace for other instances.

One or the other - whichever is appropriate. It was Identity Server down in the Q and A section, so I wasn't sure.

## Redirect Authentication

Redirect authentication is the process of delegating authentication to your Okta org and externalizes this process from the client application. The user or system is redirected to Okta for credential verification by the Identity engine, and is then provided authenticated access to the client application and other service providers.

During sign-in to the client application, users are redirected to Okta using a protocol like SAML or OIDC. After the user signs in (based on policies that are configured in Okta), Okta redirects the user back to your application. The hosted Okta Sign-in Widget can also be customized, including the domain, to match your application's brand. 

> Single Sign-on (SSO) is supported for this implementation. That is, Okta may create a session (based on the Okta policies, for example), and other integrated applications use SSO to sign in users with no configurations.

Author: As I understand it, yes, it's optional; this bit was added directly by one of the architects, so, will leave as-is. Good point about "no configuration", which could be misinterpreted. I think it works by removing that bit.

This authentication model is easy to implement, secure, and requires minimal maintenance on the part of the client application. It serves most client deployment requirements.

> An example of this flow is an organization using Okta as its Identity Provider. A user signs into an organization's on-site or cloud-based application (for example, email) and is redirected to the corporate Identity Provider, Okta, to provide sign-in and authentication. An authenticated user is then provided with a token or assertion, which provides the user access to their original application. With this existing Okta session, using SSO, the user is automatically logged into any other of the organization's service provider applications (CRM, IT, HR, and so on).

Author: Many thanks, reads much better.

> An example of this flow is when an organization uses Okta as its Identity Provider. A user tries to access the organization's on-site or cloud-based application (for example, email) and is redirected to the corporate Identity Provider, Okta, to provide sign in and authentication. After the user is authenticated, Okta provides a token or assertion to the original application to grant the user access. Okta also creates an Okta session for the authenticated user. Using SSO with this existing Okta session, the user is automatically signed in to any other of the organization's service provider applications (CRM, IT, HR, and so on).

Author: Did a search and replace for all instances. Thanks.

The interactions and communication with the Identity Provider are as follows:

<!--
See http://www.plantuml.com/plantuml/uml/

Source code for fed-auth-seq-flow.png:

@startuml
skinparam monochrome true

participant "User" as us
participant "Redirected Client" as cl
participant "Identity Server" as is

us -> cl: Resource Owner (User)
cl -> is: Redirect to IdP with Authentication Request
is -> cl: Return HTML forms to challenge user for authentication, enrollment, and so on
cl <-> us: User interacts
cl -> is: Complete Authentication Request
is -> cl: Return Assertion to Client
@enduml

 -->
![Redirect Authentication](/img/fed-auth-seq-flow.png "Redirect Authentication sequence flow")

Deploy the redirect authentication model in the following use-cases:

- When you have multiple applications, or use third- party applications and need SSO.
- You want Okta to control the authentication flows through policy configuration.
- You want Okta to control upgrades to take advantage of new functionality.
- Your application already uses an OAuth or SAML provider to sign in.

## Embedded Authentication

> Embedded authentication is the process of authenticating user credentials directly at the client application site using proprietary Okta authentication protocols. No redirection to Okta is required. The client application's code determines the methods and processes necessary to authenticate, and uses Okta’s IDX SDK to validate the credentials. Client applications create their own application sessions, and require explicit support for SSO to other external service providers.

Author: That's a good question. I'm hesitant to change this as it was provided and reviewed by the architects, but I think you're right we should modify this bit. I'll reach out to them for confirmation. Thanks.

> Embedded authentication is the process of authenticating user credentials directly at the client application site using proprietary Okta authentication protocols. No redirection to Okta is required. The client application's code determines the methods and processes necessary to authenticate, and uses Okta’s IDX SDK to validate the credentials. Client applications create their own application sessions, and require explicit support for SSO to other external service providers.

Author: Confirmed with Jeff Taylor to replace "protocols" with "SDKs and APIs"

> Embedded authentication is the process of authenticating user credentials directly at the client application site using proprietary Okta authentication protocols. No redirection to Okta is required. The client application's code determines the methods and processes necessary to authenticate, and uses Okta’s IDX SDK to validate the credentials. Client applications create their own application sessions, and require explicit support for SSO to other external service providers.

Reviewer: ```suggestion
Embedded authentication is the process of authenticating user credentials directly at the client application site using proprietary Okta authentication protocols. No redirection to Okta is required. The client application's code determines the methods and processes necessary to authenticate, and uses Okta’s Identity Experience SDK to validate the credentials. Client applications create their own application sessions and require explicit support for SSO to other external Service Providers.
```

> With embedded authentication, the user’s credentials are directly obtained by the client application and validated by API calls to Okta, generally through the customer-hosted Sign-in Widget. Based on the organization’s requirements, the client application then creates a session for user access. The client application may also exchange tokens with a Security Token Service (STS) to provide SSO access to other service providers (CRM, IT, HR, and so on). Using this deployment model, the client’s sign-in page can render native user experiences and use native platform APIs.

Reviewer 2: ```suggestion
With embedded authentication, the user’s credentials are directly obtained by the client application and validated by API calls to Okta, generally through the customer-hosted Sign-In Widget. Based on the organization’s requirements, the client application then creates a session for user access. The client application may also exchange tokens with a Security Token Service (STS) to provide SSO access to other service providers (CRM, IT, HR, and so on). Using this deployment model, the client’s sign-in page can render native user experiences and use native platform APIs.
```

> With embedded authentication, the user’s credentials are directly obtained by the client application and validated by API calls to Okta, generally through the customer-hosted Sign-In Widget. Based on the organization’s requirements, the client application then creates a session for user access. The client application may also exchange tokens with a Security Token Service (STS) to provide SSO access to other service providers (CRM, IT, HR, and so on). Using this deployment model, the client’s sign-in page can render native user experiences and use native platform APIs.

Reviewer: ```suggestion
With embedded authentication, the client application obtains and validates the user's credentials by API calls to Okta, generally through the customer-hosted Sign-In Widget. Based on the organization’s requirements, the client application then creates a session for user access. The client application may also exchange tokens with a Security Token Service (STS) to provide SSO access to other service providers (CRM, IT, HR, and so on). Using this deployment model, the client’s sign-in page can render native user experiences and use native platform APIs.
```

Author: Nice, reads better. Thanks.

> This deployment model does require more effort to implement and includes increased risk. However, it provides a deeper scope of customization and keeps users in the application with reduced redirects to Okta.

Reviewer 2: ```suggestion
This deployment model requires more effort to implement and includes increased risk. However, it does provide a deeper scope of customization and keeps users in the application with reduced redirects to Okta.
```

The detailed interactions at the client level between a client’s authorization server and resource owner are provided below:

<!--
See http://www.plantuml.com/plantuml/uml/

Source code for direct-auth-seq-flow.png:

@startuml
skinparam monochrome true

participant "Resource Owner (User)" as rs
participant "IDX Application Client" as cl
participant "Identity Server" as as

rs -> cl: Resource Owner
cl -> as: Resource Owner Password Credentials
as -> cl: Interaction required
rs <-> cl: User interacts
cl <-> as: User interacts
as -> cl: Interaction Code
cl -> as: Interaction Code
as -> cl: Tokens
@enduml

 -->

> ![Embedded Authentication](/img/direct-auth-seq-flow.png "Embedded Authentication sequence flow diagram")

Reviewer 2: ```suggestion
![Embedded authentication](/img/direct-auth-seq-flow.png "Embedded authentication sequence flow diagram")
```

## Comparison between deployment models

> The type of deployment model or authentication is dependent on your individual implementation needs and client application. In general, the method of delegating user interaction (Redirect authentication) is generally preferred for many reasons spanning security to user experience.

Reviewer 2: ```suggestion
The deployment model or authentication approach is dependent on your implementation requirements and client application. In general, the method of delegating user sign-in interaction (redirect authentication) is generally preferred for many reasons spanning security to user experience.
```

> Use this table and following Q&A to understand the differences between Redirect Authentication and Embedded Authentication, and what flow works best for your application implementation:

Reviewer 2: ```suggestion
Use this table and following Q&A to understand the differences between redirect authentication and embedded authentication, and what flow works best for your application implementation:
```

> Use this table and following Q&A to understand the differences between redirect authentication and embedded authentication, and what flow works best for your application implementation:

Reviewer: ```suggestion
Use this table and the following Q&A to understand the differences between redirect authentication and embedded authentication, and what flow works best for your application implementation:
```

> | Change                                                    | Redirect Authentication | Embedded Authentication |

Reviewer 2: "Change"?
Redirect Authentication => Redirect authentication
Embedded Authentication => Embedded authentication

| --------------------------------------------------------- | ------------------------ | --------------------- |
| Client application requires a redirect to an Identity Provider.   |   &#9989;     | |

> | Client application tightly coupled with authentication methods and processes.  |        | &#9989; |

Reviewer 2: ```suggestion
| Client application is tightly coupled with authentication methods and processes.  |        | &#9989; |
```

> | SSO is implicit (if session created, SSO implemented for other resources)   |   &#9989;     | |

Reviewer 2: ```suggestion
| SSO is implicit (if Okta session is created, SSO is implemented for other resources)   |   &#9989;     | |
```

Reviewer 2: Not sure if the session here is referring to the Okta session or the application session? Please confirm.

Author: I believe it's an Okta session, but will need to confirm. Thanks.

Author: Confirmed with Jeff Taylor this is Okta session.

> | Application customization is a key requirement   |        | &#9989;|

Reviewer 2: ```suggestion
| Application customization is a key requirement.   |        | &#9989;|
```

Further questions to determine what authentication flow works best for your implementation:

Q. Does the application want to own the authentication and registration process without leaving the application?

> A. Embedded Authentication

Reviewer 2: ```suggestion
A. embedded authentication
```

Q. Does the application want to own user remediation (communication with Identity Server)?

> A. Embedded Authentication

Reviewer 2: ```suggestion
A. embedded authentication
```

> Q. Does the application need to support external identity providers, authenticators,  or claims providers?

Reviewer 2: ```suggestion
Q. Does the application need to support external identity providers, authenticators, or claims providers?
```

> A. Redirect Authentication or Embedded Authentication

Reviewer 2: ```suggestion
A. redirect authentication or embedded authentication
```

> Q. Does the application need to support external identity providers, authenticators, or claims providers?

Reviewer: ```suggestion
Q. Does the application need to support external Identity Providers, authenticators, or claims providers?
```

> Q. Does the application need to SSO to downstream resources?

Reviewer 2: ```suggestion
Q. Does the application need SSO for downstream resources?
```

> A. Redirect Authentication

Reviewer 2: ```suggestion
A. redirect authentication
```

Q. Does the application need centralized session management across applications?

> A. Redirect Authentication

Reviewer 2: ```suggestion
A. redirect authentication
```
